[extractor/common] detect media playlist in _extract_m3u8_formats

remitamine opened the pull request:
fixes the broblem with smil manifests that contain m3u8 media playlists:
```
python2 __main__.py -F 'http://link.theplatform.com/s/dJ5BDC/eSDhwOPqdCjD?format=SMIL&mbr=true'
[ThePlatform] eSDhwOPqdCjD: Downloading SMIL data
[ThePlatform] eSDhwOPqdCjD: Downloading m3u8 information
[ThePlatform] eSDhwOPqdCjD: Downloading m3u8 information
[ThePlatform] eSDhwOPqdCjD: Downloading m3u8 information
[ThePlatform] eSDhwOPqdCjD: Downloading m3u8 information
[ThePlatform] eSDhwOPqdCjD: Downloading m3u8 information
[ThePlatform] eSDhwOPqdCjD: Downloading m3u8 information
[ThePlatform] eSDhwOPqdCjD: Downloading m3u8 information
[ThePlatform] eSDhwOPqdCjD: Downloading JSON metadata
[info] Available formats for eSDhwOPqdCjD:
format code  extension  resolution note
hls-meta-0   mp4        multiple   Quality selection URL 
hls-meta-1   mp4        multiple   Quality selection URL 
hls-meta-2   mp4        multiple   Quality selection URL 
hls-meta-3   mp4        multiple   Quality selection URL 
hls-meta-4   mp4        multiple   Quality selection URL 
hls-meta-5   mp4        multiple   Quality selection URL 
hls-meta-6   mp4        multiple   Quality selection URL 
7            ts         unknown    
8            ts         unknown    
9            ts         unknown    
10           ts         unknown    
...
550          aac        unknown    
551          aac        unknown    
552          aac        unknown    
553          aac        unknown    
554          aac        unknown    
555          aac        unknown    
556          aac        unknown    
557          aac        unknown    
558          aac        unknown    
559          aac        unknown    (best)
```
with the change:
```
python2 __main__.py -F 'http://link.theplatform.com/s/dJ5BDC/eSDhwOPqdCjD?format=SMIL&mbr=true'
[ThePlatform] eSDhwOPqdCjD: Downloading SMIL data
[ThePlatform] eSDhwOPqdCjD: Downloading m3u8 information
[ThePlatform] eSDhwOPqdCjD: Downloading m3u8 information
[ThePlatform] eSDhwOPqdCjD: Downloading m3u8 information
[ThePlatform] eSDhwOPqdCjD: Downloading m3u8 information
[ThePlatform] eSDhwOPqdCjD: Downloading m3u8 information
[ThePlatform] eSDhwOPqdCjD: Downloading m3u8 information
[ThePlatform] eSDhwOPqdCjD: Downloading m3u8 information
[ThePlatform] eSDhwOPqdCjD: Downloading JSON metadata
[info] Available formats for eSDhwOPqdCjD:
format code  extension  resolution note
hls-1        m3u8       400x224       1k 
hls-110      m3u8       400x224     110k 
hls-250      m3u8       400x224     250k 
hls-500      m3u8       512x288     500k 
hls-800      m3u8       640x360     800k 
hls-1200     m3u8       768x432    1200k 
hls-1800     m3u8       960x540    1800k  (best)
```

dstftw commented on youtube_dl/extractor/common.py:
```diff
@@ -1014,6 +1014,12 @@ def _extract_m3u8_formats(self, m3u8_url, video_id, ext=None,
             return []
         m3u8_doc, urlh = res
         m3u8_url = urlh.geturl()
+        is_media_playlist = re.search(r'#EXT-X-TARGETDURATION', m3u8_doc)
```
This heuristic should be clarified with rationale comment. It can also be simplified to: 
```python
if '#EXT-X-TARGETDURATION' in m3u8_doc:
```
It's also probably makes sense to additionaly check for missing `#EXT-X-STREAM-INF`.

remitamine commented:
> This heuristic should be clarified with rationale comment. It can also be simplified to:

i added a comment and simplified the condition as you suggest.

> It's also probably makes sense to additionaly check for missing #EXT-X-STREAM-INF.

is it required for every Master Playlist(i didn't something about it in https://tools.ietf.org/html/draft-pantos-http-live-streaming-17).

dstftw commented:
It's generally a precaution for possibly malformed playlists.

remitamine commented:
if #EXT-X-STREAM-INF is missing shoud i report a warning or just return an empty array?

dstftw commented:
I mean:
```python
if '#EXT-X-TARGETDURATION' in m3u8_doc and '#EXT-X-STREAM-INF' not in m3u8_doc:
    ...
```
But ok let's just keep it as is until this kind of malformed playlist is seen (if ever).

remitamine commented:
can i merge it now?

dstftw commented on youtube_dl/extractor/common.py:
```diff
@@ -1203,8 +1216,18 @@ def _parse_smil_formats(self, smil, smil_url, video_id, namespace=None, f4m_para
+                m3u8_formats = self._extract_m3u8_formats(
+                    src_url, video_id, ext or 'mp4', m3u8_id='hls', fatal=False)
+                if m3u8_formats:
+                    if len(m3u8_formats) == 1:
```
This can be simplified to just `if len(m3u8_formats) == 1:`.

remitamine replied:
it's an old habit(i forget that the _extract_m3u8_formats changed to always retrun an array).
i changed it now.

dstftw commented:
LGTM.

remitamine commented:
now i can fix the cbsnews extraction.
i will open another PR to add ability to customize the theplatform metadata extract(needed to fix this problem https://github.com/rg3/youtube-dl/commit/220ee33f2b8a22d2fe992c26b32a046b77ce2a4a#commitcomment-15629846).
